mixer/adapter/svcctrl/lrucache.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package svcctrl
+
+import (
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru/simplelru"
+	"github.com/jonboulle/clockwork"
+)
+
+type lruCache interface {
+	add(key, value interface{}) bool
+	get(key interface{}) (interface{}, bool)
+	remove(key interface{}) bool
+	purge()
+}
+
+type lruCacheValue struct {
+	// insert timestamp
+	timestamp time.Time
+	value     interface{}
+}
+
+type cache struct {
+	mutex      sync.RWMutex
+	lru        *simplelru.LRU
+	clock      clockwork.Clock
+	expiration time.Duration
+}
+
+func (c *cache) add(key, value interface{}) bool {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	return c.lru.Add(key, &lruCacheValue{
+		c.clock.Now(),
+		value,
+	})
+}
+
+func (c *cache) get(key interface{}) (interface{}, bool) {
+	value := c.getImpl(key)
+	if value == nil {
+		return nil, false
+	}
+
+	if c.clock.Since(value.timestamp) > c.expiration {
+		c.mutex.Lock()
+		defer c.mutex.Unlock()
+		tmp, ok := c.lru.Get(key)
+		if ok {
+			valueInCache := tmp.(*lruCacheValue)
+			if value == valueInCache ||
+				c.clock.Since(valueInCache.timestamp) > c.expiration {
+				c.lru.Remove(key)
+				return nil, false
+			}
+			return valueInCache.value, true
+		}
+	}
+	return value.value, true
+}
+
+func (c *cache) remove(key interface{}) bool {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	return c.lru.Remove(key)
+}
+
+func (c *cache) purge() {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.lru.Purge()
+}
+
+func (c *cache) getImpl(key interface{}) *lruCacheValue {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	value, ok := c.lru.Get(key)
+	if !ok {
+		return nil
+	}
+	return value.(*lruCacheValue)
+}
+
+func newLruCache(size int, expiration time.Duration) (lruCache, error) {
```

Comment:
nit: golang convention is to make all initialisms caps... so `newLRUCache` instead of `newLruCache`.  https://github.com/golang/go/wiki/CodeReviewComments#initialisms

Reply:
Done